syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.env";
option java_outer_classname = "ReplayBufferProto";
option objc_class_prefix = "HLW";

package actor;

// The greeting service definition.
service Actor {
  rpc ServiceReady (ServiceReadyReq) returns (ServiceReadyResp) {}
  rpc Start (stream StartReq) returns (StartResp) {}
  rpc ReplicateModel (stream ReplicateModelReq) returns (ReplicateModelResp) {}
  rpc Terminate (TerminateReq) returns (TerminateResp) {}
}

message ServiceReadyReq {
  string requestId = 1;
}

message ServiceReadyResp {
  bool ready = 1;
  string requestId = 2;
}

message StartMeta {
  string filename = 1;
  string trainerType = 2;
  int64 maxEpisode = 3;
  int64 maxEpisodeStep = 4;
  string actionTransform = 5;
  uint64 version = 6;
  string requestId = 7;
}

message StartReq {
  oneof request {
    StartMeta meta = 1;
    bytes chunkData = 2;
  }
}

message StartResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}

message ReplicateModelMeta {
  string filename = 1;
  uint64 version = 2;
  string requestId = 3;
}

message ReplicateModelReq {
  oneof request {
    ReplicateModelMeta meta = 1;
    bytes chunkData = 2;
  }
}

message ReplicateModelResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}

message TerminateReq {
  string requestId = 1;
}

message TerminateResp {
  bool done = 1;
  string err = 2;
  string requestId = 3;
}
